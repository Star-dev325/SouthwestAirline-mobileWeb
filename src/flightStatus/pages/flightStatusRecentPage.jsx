// @flow
import React from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';
import withConnectedReactRouter from 'src/shared/enhancers/withConnectedReactRouter';
import EditRecentSearches from 'src/shared/components/editRecentSearches';
import FlightStatusRecentSearchCard from 'src/flightStatus/components/flightStatusRecentSearchCard';
import * as FlightStatusSearchActions from 'src/flightStatus/actions/flightStatusSearchActions';
import * as FormDataActions from 'src/shared/actions/formDataActions';
import { FLIGHT_STATUS_SEARCH_FORM } from 'src/shared/constants/formIds';
import type { RecentSearchRequestType } from 'src/flightStatus/flow-typed/flightStatus.types';
import type { GoBack } from 'src/shared/flow-typed/shared.types';

type Props = {
  goBack: GoBack,
  saveSelectedRecentSearchRequestFn: (RecentSearchRequestType) => void,
  clearFormDataByIdFn: (string) => void,
  onDeleteCurrentSearchFn: (RecentSearchRequestType) => void,
  searches: Array<RecentSearchRequestType>
};

export class FlightStatusRecentPage extends React.Component<Props> {
  _onRecentSearchCardClicked = (searchRequest: RecentSearchRequestType) => {
    const { saveSelectedRecentSearchRequestFn, clearFormDataByIdFn, goBack } = this.props;

    saveSelectedRecentSearchRequestFn(searchRequest);
    clearFormDataByIdFn(FLIGHT_STATUS_SEARCH_FORM);
    goBack();
  };

  render() {
    return (
      <EditRecentSearches
        listOfRecentSearches={this.props.searches}
        recentSearchComponent={FlightStatusRecentSearchCard}
        onDeleteCurrentSearch={this.props.onDeleteCurrentSearchFn}
        onRecentSearchCardClicked={this._onRecentSearchCardClicked}
      />
    );
  }
}

const mapStateToProps = (state) => ({
  searches: _.get(state, 'app.flightStatus.flightStatusRecentPage.searches')
});

const mapDispatchToProps = {
  saveSelectedRecentSearchRequestFn: FlightStatusSearchActions.saveSelectedRecentSearchRequest,
  clearFormDataByIdFn: FormDataActions.clearFormDataById,
  onDeleteCurrentSearchFn: FlightStatusSearchActions.deleteRecentSearchRequest
};

const enhancers = _.flowRight(withConnectedReactRouter, connect(mapStateToProps, mapDispatchToProps));

export default enhancers(FlightStatusRecentPage);
