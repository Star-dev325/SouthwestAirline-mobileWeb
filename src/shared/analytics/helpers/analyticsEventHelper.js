import _ from 'lodash';

const pageIdentifierTransformer = [
  { matcher: /^\/$|^\/\?/, id: 'home' },
  { matcher: /\/flight-status\/(?!.*html).*/, id: 'flight-status-results' },

  { matcher: /\/flight-status\?.*_modal=originAirport/, id: 'flight-status-from' },
  { matcher: /\/flight-status\?.*_modal=destinationAirport/, id: 'flight-status-to' },

  { matcher: /\/air\/booking\/shopping\?.*_modal=from/, id: 'air-booking-shopping-from' },
  { matcher: /\/air\/booking\/shopping\?.*_modal=to/, id: 'air-booking-shopping-to' },
  { matcher: /\/air\/booking\/shopping\?.*_modal=airBookingCalendar/, id: 'air-booking-shopping-dates' },

  { matcher: /\/air\/change\/shopping\?.*_modal=from/, id: 'air-change-shopping-from' },
  { matcher: /\/air\/change\/shopping\?.*_modal=to/, id: 'air-change-shopping-to' },
  { matcher: /\/air\/change\/shopping\?.*_modal=airBookingCalendar/, id: 'air-change-shopping-dates' },

  { matcher: /\/car\/booking\?.*_modal=pickUpModal/, id: 'car-booking-pick-up' },
  { matcher: /\/car\/booking\?.*_modal=dropOffModal/, id: 'car-booking-drop-off' },
  { matcher: /\/car\/booking\?.*_modal=carBookingCalendar/, id: 'car-booking-dates' },

  { matcher: /\/check-in\/reservation\/\w{6}$/, id: 'checkin-view-reservation' }
];

const _matchPath = (path) =>
  _.get(
    _.find(pageIdentifierTransformer, (t) => path.match(t.matcher)),
    'id'
  );

export const transformPath = (path, search = null) => {
  if (!path) {
    return '';
  }
  let toMatchPath = path;

  if (search) {
    toMatchPath = `${path}${search}`;
  }

  return _matchPath(toMatchPath) || path.replace(/\//g, '-').slice(1);
};

export const raiseEvent = (eventName) => {
  window._trackAnalytics && window._trackAnalytics.event();

  const customEvent = new window.CustomEvent(eventName);

  document.body.dispatchEvent(customEvent);
};

export const raiseSatelliteEvent = (eventName, detail) => {
  if (window._satellite) {
    detail ? window._satellite.track(eventName, detail) : window._satellite.track(eventName);
  }
};

export const customerMessageAnalytics = (messages) => {
  if (messages) {
    const messageKeys = _.chain(messages).map('key').compact().join().value();

    _.set(window, 'data_a.message.customer', messageKeys);
    _.set(window, 'data_a.message.customerdisplay', '1');
  }
};

export const getPageIdentifier = (location) => {
  const { pathname = '', search = '' } = location || {};

  return transformPath(pathname, search);
};
