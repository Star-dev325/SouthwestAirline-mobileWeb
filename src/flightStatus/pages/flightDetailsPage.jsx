// @flow

import i18n from '@swa-ui/locale';
import cx from 'classnames';
import _ from 'lodash';
import React from 'react';
import { connect } from 'react-redux';
import * as FlightStatusSearchActions from 'src/flightStatus/actions/flightStatusSearchActions';
import FlightStatusDetailCard from 'src/flightStatus/components/flightStatusDetailCard';
import { shareFlightStatusDetails } from 'src/shared/actions/webViewActions';
import { raiseSatelliteEvent } from 'src/shared/analytics/helpers/analyticsEventHelper';
import AircraftTypeFooter from 'src/shared/components/aircraftTypeFooter';
import Container from 'src/shared/components/container';
import PageHeader from 'src/shared/components/pageHeader';
import SearchFlightsSummaryHeader from 'src/shared/components/searchFlightsSummaryHeader';
import withBodyClass from 'src/shared/enhancers/withBodyClass';
import withConnectedReactRouter from 'src/shared/enhancers/withConnectedReactRouter';

import type { Push } from 'src/shared/flow-typed/shared.types';
import type {
  FlightSearchType,
  FlightStatusDetailsPageType,
  RecentSearchRequestType
} from 'src/flightStatus/flow-typed/flightStatus.types';

type Props = {
  AIRCRAFT_TYPE_FLIGHTSTATUS?: boolean,
  fetchFlightDetailsFn: (RecentSearchRequestType, boolean, ?Push, boolean, boolean) => *,
  flightStatusDetailsPage: {
    response: ?FlightStatusDetailsPageType
  },
  isWebView: boolean,
  query: FlightSearchType,
  shareFlightStatus: boolean,
  shareFlightStatusDetailsFn: (*) => void
};

export class FlightDetailsPage extends React.Component<Props> {
  componentDidMount() {
    if (_.isEmpty(this.props.flightStatusDetailsPage.response)) {
      const { fetchFlightDetailsFn, query } = this.props;
      const withFlightKeys = query?.flightKeys;

      fetchFlightDetailsFn(query, false, null, true, !!withFlightKeys);
    }
  }

  _onRefreshClicked = () => {
    const { fetchFlightDetailsFn, query } = this.props;
    const withFlightKeys = query?.flightKeys;

    fetchFlightDetailsFn(query, false, null, true, !!withFlightKeys);
  };

  _onShareClicked = () => {
    const { shareFlightStatusDetailsFn, flightStatusDetailsPage } = this.props;
    const shareDetails = _.get(flightStatusDetailsPage, 'response.shareDetails');

    if (shareDetails) {
      shareFlightStatusDetailsFn(shareDetails);
      raiseSatelliteEvent('share flight details');
    }
  };

  renderRefreshButton(isWebView: boolean) {
    return (
      isWebView && (
        <a className="right white normal page-header--right-button" onClick={this._onRefreshClicked}>
          {i18n('SHARED__FLIGHT_STATUS__REFRESH')}
        </a>
      )
    );
  }

  render() {
    if (_.isEmpty(this.props.flightStatusDetailsPage.response)) {
      return null;
    }

    const {
      AIRCRAFT_TYPE_FLIGHTSTATUS,
      flightStatusDetailsPage: { response },
      isWebView,
      shareFlightStatus
    } = this.props;
    const tripDescription = _.get(response, 'header.tripDescription', '');
    const from = _.get(response, 'header.from', '');
    const to = _.get(response, 'header.to', '');
    const date = _.get(response, 'header.date', '');
    const flightCards = _.get(response, 'flightCards', []);

    return (
      <div className="flight-details">
        <PageHeader noBottomPadding>
          <span className={cx({ 'page-title': !isWebView }, { hide: isWebView })}>Flight Status</span>
          <span className={cx('page-title-details', { m0: isWebView })}>{tripDescription}</span>

          {this.renderRefreshButton(isWebView)}
        </PageHeader>

        <div className="flight-details-content">
          <SearchFlightsSummaryHeader
            date={date}
            from={from}
            to={to}
            onShareClickedCb={shareFlightStatus && isWebView ? this._onShareClicked : null}
          />
          <Container className="results">
            {flightCards.map((flightCard, id) => (
              <FlightStatusDetailCard
                key={id}
                flightCard={flightCard}
                AIRCRAFT_TYPE_FLIGHTSTATUS={AIRCRAFT_TYPE_FLIGHTSTATUS}
              />
            ))}
          </Container>
          {AIRCRAFT_TYPE_FLIGHTSTATUS && <AircraftTypeFooter />}
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  flightStatusDetailsPage: _.get(state, 'app.flightStatus.flightStatusDetailsPage'),
  AIRCRAFT_TYPE_FLIGHTSTATUS: _.get(state, 'app.toggles.AIRCRAFT_TYPE_FLIGHTSTATUS'),
  shareFlightStatus: _.get(state, 'app.webView.shareFlightStatus'),
  isWebView: _.get(state, 'app.webView.isWebView')
});

const mapDispatchToProps = {
  fetchFlightDetailsFn: FlightStatusSearchActions.fetchFlightDetails,
  shareFlightStatusDetailsFn: shareFlightStatusDetails
};

const enhancers = _.flowRight(
  withConnectedReactRouter,
  withBodyClass('flight-details-bg'),
  connect(mapStateToProps, mapDispatchToProps)
);

export default enhancers(FlightDetailsPage);
