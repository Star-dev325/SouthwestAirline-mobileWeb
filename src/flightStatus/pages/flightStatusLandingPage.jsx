// @flow

import i18n from '@swa-ui/locale';
import cx from 'classnames';
import _ from 'lodash';
import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import * as AirportInfoActions from 'src/airports/actions/airportInfoActions';
import * as AirportsActions from 'src/airports/actions/airportsActions';
import * as FlightStatusSearchActions from 'src/flightStatus/actions/flightStatusSearchActions';
import FlightStatusSearchForm from 'src/flightStatus/components/flightStatusSearchForm';
import { transformToFlightSearchRequest } from 'src/flightStatus/transformers/flightStatusTransformer';
import * as FormDataActions from 'src/shared/actions/formDataActions';
import * as AnalyticsActions from 'src/shared/analytics/actions/analyticsActions';
import Container from 'src/shared/components/container';
import PageHeader from 'src/shared/components/pageHeader';
import { FLIGHT_STATUS_SEARCH_FORM } from 'src/shared/constants/formIds';
import withBodyClass from 'src/shared/enhancers/withBodyClass';
import withConnectedReactRouter from 'src/shared/enhancers/withConnectedReactRouter';
import withFeatureToggles from 'src/shared/enhancers/withFeatureToggles';
import { getNormalizedRoute } from 'src/shared/helpers/urlHelper';

import type { RecentSearchRequestType, SearchFormData } from 'src/flightStatus/flow-typed/flightStatus.types';
import type { AirportType, Push } from 'src/shared/flow-typed/shared.types';

type Props = {
  allAirports: Array<AirportType>,
  analyticsTrackSubmitFormFn: (*) => void,
  fetchFlightDetailsFn: (RecentSearchRequestType, boolean, Push, boolean) => *,
  fetchFlightStatusFn: (RecentSearchRequestType, boolean, Push) => *,
  getRecentSearchesFromLocalStorageFn: () => void,
  isWebView: boolean,
  loadAirportsFn: () => void,
  loadRecentlySearchedFn: () => void,
  push: Push,
  query?: {
    departureDate?: string,
    destinationAirportCode?: string,
    originationAirportCode?: string,
    flightNumber?: string
  },
  recentlySearched: Array<AirportType>,
  selectedRecentSearchRequest: ?RecentSearchRequestType,
  updateFormDataValueFn: (string, *) => void,
  updateSelectedAirportInfoFn: (airportInfo: *) => void
};

export class FlightStatusLandingPage extends React.Component<Props> {
  componentDidMount() {
    const { getRecentSearchesFromLocalStorageFn, loadAirportsFn, loadRecentlySearchedFn, query, updateFormDataValueFn } = this.props;

    if (query)  {
      const { departureDate, destinationAirportCode, flightNumber, originationAirportCode } = query;

      updateFormDataValueFn(FLIGHT_STATUS_SEARCH_FORM, {
        destinationAirport: destinationAirportCode,
        flightNumber,
        originAirport: originationAirportCode,
        selectedDate: departureDate
      });
    }

    loadAirportsFn();
    loadRecentlySearchedFn();
    getRecentSearchesFromLocalStorageFn();
  }

  _onSubmit = (formData: SearchFormData) => {
    const { analyticsTrackSubmitFormFn, updateFormDataValueFn } = this.props;
    const searchRequest = transformToFlightSearchRequest(
      formData.originAirport,
      formData.destinationAirport,
      formData.selectedDate,
      formData.flightNumber
    );

    updateFormDataValueFn(FLIGHT_STATUS_SEARCH_FORM, {
      originAirport: formData.originAirport,
      destinationAirport: formData.destinationAirport,
      selectedDate: formData.selectedDate,
      flightNumber: formData.flightNumber
    });
    analyticsTrackSubmitFormFn('flight-status-search');
    this._transitionToNextPage(searchRequest);
  };

  _transitionToNextPage(searchRequest: RecentSearchRequestType) {
    const { fetchFlightStatusFn, fetchFlightDetailsFn, push } = this.props;

    if (_.isEmpty(searchRequest.flightNumber)) {
      fetchFlightStatusFn(searchRequest, true, push);
    } else {
      fetchFlightDetailsFn(searchRequest, true, push, true);
    }
  }

  render() {
    let newFormData = {};
    const { selectedRecentSearchRequest, allAirports, recentlySearched, updateSelectedAirportInfoFn, isWebView } =
      this.props;

    if (selectedRecentSearchRequest) {
      newFormData = {
        originAirport: _.get(selectedRecentSearchRequest, 'from'),
        destinationAirport: _.get(selectedRecentSearchRequest, 'to'),
        selectedDate: _.get(selectedRecentSearchRequest, 'date')
      };

      if (!_.isEmpty(selectedRecentSearchRequest.flightNumber)) {
        _.set(newFormData, 'flightNumber', selectedRecentSearchRequest.flightNumber);
      }
    }

    return (
      <div>
        <PageHeader className={cx({ 'center caps bgsdkblue px0': isWebView })} noBottomPadding>
          {isWebView ? 'SEARCH' : i18n('SHARED__FLIGHT_STATUS__TITLE')}
          <Link
            data-qa="recent-search-button"
            className={cx(
              'right white page-header--right-button',
              { regular: !isWebView },
              { 'bold center halfwidth bgsblue': isWebView }
            )}
            to={getNormalizedRoute({ routeName: 'recent' })}
          >
            {i18n('SHARED__FLIGHT_STATUS__RECENT')}
          </Link>
        </PageHeader>
        <Container>
          <FlightStatusSearchForm
            formId={FLIGHT_STATUS_SEARCH_FORM}
            onSubmit={this._onSubmit}
            initialFormData={newFormData}
            updateSelectedAirportInfoFn={updateSelectedAirportInfoFn}
            allAirports={allAirports}
            recentlySearched={recentlySearched}
            isWebView={isWebView}
          />
        </Container>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  selectedRecentSearchRequest: _.get(state, 'app.flightStatus.selectedRecentSearchRequest'),
  allAirports: _.get(state, 'app.airports.allAirports'),
  recentlySearched: _.get(state, 'app.airports.recentlySearched'),
  isWebView: state.app.webView.isWebView
});

const mapDispatchToProps = {
  fetchFlightStatusFn: FlightStatusSearchActions.fetchFlightStatus,
  fetchFlightDetailsFn: FlightStatusSearchActions.fetchFlightDetails,
  getRecentSearchesFromLocalStorageFn: FlightStatusSearchActions.getRecentSearchesFromLocalStorage,
  analyticsTrackSubmitFormFn: AnalyticsActions.trackSubmitForm,
  updateFormDataValueFn: FormDataActions.updateFormDataValue,
  loadAirportsFn: AirportsActions.loadAirports,
  loadRecentlySearchedFn: AirportsActions.loadRecentlySearched,
  updateSelectedAirportInfoFn: AirportInfoActions.updateSelectedAirportInfo
};

const enhancers = _.flowRight(
  withBodyClass('flight-status-search'),
  withConnectedReactRouter,
  withFeatureToggles,
  connect(mapStateToProps, mapDispatchToProps)
);

export default enhancers(FlightStatusLandingPage);
