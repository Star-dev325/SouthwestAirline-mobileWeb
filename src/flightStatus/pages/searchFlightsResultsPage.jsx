// @flow
import _ from 'lodash';
import React from 'react';
import { connect } from 'react-redux';
import * as FlightStatusSearchActions from 'src/flightStatus/actions/flightStatusSearchActions';
import FlightCard from 'src/flightStatus/components/flightCard';
import type { FlightSchedulesPageType, RecentSearchRequestType } from 'src/flightStatus/flow-typed/flightStatus.types';
import Container from 'src/shared/components/container';
import PageHeader from 'src/shared/components/pageHeader';
import SearchFlightsSummaryHeader from 'src/shared/components/searchFlightsSummaryHeader';
import withBodyClass from 'src/shared/enhancers/withBodyClass';
import withConnectedReactRouter from 'src/shared/enhancers/withConnectedReactRouter';
import type { Push } from 'src/shared/flow-typed/shared.types';
import { isOnOldRoute } from 'src/shared/helpers/urlHelper';

type Props = {
  flightSchedulesPage: {
    response: FlightSchedulesPageType
  },
  isWebView?: boolean,
  fetchFlightStatusFn: (searchRequest: RecentSearchRequestType, shouldGoToNextPage: boolean, push: Push) => *,
  push: Push,
  params: {
    from: string,
    to: string,
    date: string
  },
  query?: {
    departureDate: string,
    destinationAirportCode: string,
    originationAirportCode: string
  },
  lookUpFlightStatusDetailsFn: (Link, Push) => *
};

export class SearchFlightsResultsPage extends React.Component<Props> {
  componentDidMount() {
    const { fetchFlightStatusFn, flightSchedulesPage, params, push, query } = this.props;
    
    if (_.isEmpty(flightSchedulesPage.response)) {
      let { date, from, to  } = params;

      if (!isOnOldRoute() && query)  {
        const { departureDate, destinationAirportCode, originationAirportCode } = query;
        
        date = departureDate;
        from = originationAirportCode;
        to = destinationAirportCode;
      }

      fetchFlightStatusFn({ date, from, to }, false, push);
    }
  }

  _onFlightCardClicked = (flightStatusDetail: Link) => {
    const { lookUpFlightStatusDetailsFn, push } = this.props;

    lookUpFlightStatusDetailsFn(flightStatusDetail, push);
  };

  renderBody() {
    const {
      flightSchedulesPage: { response }
    } = this.props;
    const from = _.get(response, 'header.from', '');
    const to = _.get(response, 'header.to', '');
    const date = _.get(response, 'header.date', '');
    const flights = _.get(response, 'flights', []);

    return (
      <div ref="loadedBody" className="page container" data-qa="search flight results page">
        <SearchFlightsSummaryHeader from={from} to={to} date={date} />
        <Container>
          {flights.map((flight, id) => (
            <FlightCard key={id} flight={flight} onFlightCardClicked={this._onFlightCardClicked} />
          ))}
        </Container>
      </div>
    );
  }

  render() {
    if (_.isEmpty(this.props.flightSchedulesPage.response)) {
      return null;
    }

    const {
      flightSchedulesPage: { response },
      isWebView
    } = this.props;
    const tripDescription = _.get(response, 'header.tripDescription', '');

    return (
      <div>
        <PageHeader>
          <span className="page-title">
            {isWebView ? '' : 'Flight Status'} <span className="normal italics">{tripDescription}</span>
          </span>
        </PageHeader>

        {this.renderBody()}
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  flightSchedulesPage: _.get(state, 'app.flightStatus.flightSchedulesPage'),
  isWebView: state.app.webView.isWebView
});

const mapDispatchToProps = {
  lookUpFlightStatusDetailsFn: FlightStatusSearchActions.lookUpFlightStatusDetails,
  fetchFlightStatusFn: FlightStatusSearchActions.fetchFlightStatus
};

const enhancers = _.flowRight(
  withConnectedReactRouter,
  withBodyClass('flight-status-search-result'),
  connect(mapStateToProps, mapDispatchToProps)
);

export default enhancers(SearchFlightsResultsPage);
