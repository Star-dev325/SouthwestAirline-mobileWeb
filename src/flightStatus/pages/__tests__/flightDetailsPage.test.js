import '@testing-library/jest-dom/extend-expect';
import { fireEvent } from '@testing-library/react';
import { FlightDetailsPage } from 'src/flightStatus/pages/flightDetailsPage';
import * as analyticsEventHelper from 'src/shared/analytics/helpers/analyticsEventHelper';
import FlightStatusDetailsBuilder from 'test/builders/apiResponse/v1/mobile-air-operations/page/flight-status/flightStatusDetailsBuilder';
import { createComponent } from 'test/unit/helpers/testingLibraryUtils';
import waitFor from 'test/unit/helpers/waitFor';

describe('Flight Details Page', () => {
  let fetchFlightDetailsFnMock;
  let shareFlightStatusDetailsFnMock;

  beforeEach(() => {
    fetchFlightDetailsFnMock = jest.fn();
    shareFlightStatusDetailsFnMock = jest.fn();
    jest.spyOn(analyticsEventHelper, 'raiseSatelliteEvent');
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  describe('render', () => {
    it('renders those flight details', (done) => {
      const { container } = createPage();

      waitFor.untilAssertPass(() => {
        expect(container.querySelector('.flight-details')).toBeInTheDocument();
        expect(container.querySelector('.page-header--right-button')).not.toBeInTheDocument();
      }, done);
    });

    describe('render', () => {
      it('should render single leg flight when flightStatusDetailsPage is not empty', (done) => {
        const { container } = createPage();

        waitFor.untilAssertPass(() => {
          expect(container.querySelectorAll('.segment-details')).toHaveLength(1);
          expect(container.querySelectorAll('.leg-details')).toHaveLength(1);
          expect(container.querySelector('.page-header--right-button')).not.toBeInTheDocument();
        }, done);
      });

      it('should render multi leg flight when flightStatusDetailsPage is not empty', () => {
        const FlightStatusDetailsPageResponse = new FlightStatusDetailsBuilder()
          .withConnectingFlight()
          .build().flightStatusDetailsPage;

        const props = {
          query: {
            from: 'ATL',
            to: 'HOU',
            date: '2017-04-09',
            flightNumber: '100',
            connectingAirportCode: 'MSY',
            secondFlightNumber: '205'
          },
          flightStatusDetailsPage: {
            response: FlightStatusDetailsPageResponse
          }
        };
        const { container } = createPage(props);

        expect(container.querySelectorAll('.segment-details')).toHaveLength(1);
        expect(container.querySelectorAll('.leg-details')).toHaveLength(2);
        expect(container.querySelector('.page-header--right-button')).not.toBeInTheDocument();
      });

      it('should render multi leg flight when flightStatusDetailsPage is not empty with flight-keys', () => {
        const FlightStatusDetailsPageResponse = new FlightStatusDetailsBuilder()
          .withConnectingFlight()
          .build().flightStatusDetailsPage;

        const props = {
          query: {
            "flight-keys": "2023-11-10:AUSHOU1228|2023-11-10:HOUDAL20"
          },
          flightStatusDetailsPage: {
            response: FlightStatusDetailsPageResponse
          }
        };
        const { container } = createPage(props);

        expect(container.querySelectorAll('.segment-details')).toHaveLength(1);
        expect(container.querySelectorAll('.leg-details')).toHaveLength(2);
        expect(container.querySelector('.page-header--right-button')).not.toBeInTheDocument();
      });

      it('should not render when flightStatusDetailsPage is empty', () => {
        const { container } = createPage({ flightStatusDetailsPage: { response: null } });

        expect(container.querySelector('.segment-details')).not.toBeInTheDocument();
        expect(container.querySelector('.leg-details')).not.toBeInTheDocument();
        expect(container.querySelector('.page-header--right-button')).not.toBeInTheDocument();
      });

      describe('in webView', () => {
        it('and should render refresh button', (done) => {
          const { container } = createPage({ isWebView: true });

          waitFor.untilAssertPass(() => {
            expect(container.querySelector('.page-header--right-button')).toBeInTheDocument();
          }, done);
        });

        it('and should render Share button when shareFlightStatus is true', (done) => {
          const { container } = createPage({ isWebView: true, shareFlightStatus: true });

          waitFor.untilAssertPass(() => {
            expect(container.querySelector('.share-button')).toBeInTheDocument();
          }, done);
        });
      });
    });
  });

  describe('trip description', () => {
    it('should should display page header use flight numbers', () => {
      const { container } = createPage();

      expect(container.querySelector('.page-title-details').textContent).toEqual('Flight 1632 / 1781');
    });
  });

  it('should pass correct date, from and to props to component SearchFlightsSummaryHeader', () => {
    const { container } = createPage();

    expect(container).toMatchSnapshot();
  });

  it('should map multiple FlightStatusDetailCard use flightCards', (done) => {
    const FlightStatusDetailsPageResponse = new FlightStatusDetailsBuilder().build().flightStatusDetailsPage;
    const props = {
      flightStatusDetailsPage: {
        response: FlightStatusDetailsPageResponse
      }
    };
    const { container } = createPage(props);

    waitFor.untilAssertPass(() => {
      expect(container.querySelectorAll('.segment-details')).toHaveLength(1);
    }, done);
  });

  describe('fetchFlightDetailsFn called', () => {
    it('should call fetchFlightDetailsFnStub when component mounts', () => {
      createPage({}, true);
      expect(fetchFlightDetailsFnMock).toBeCalled();
    });
  });

  describe('in WebView and shareFlightStatusDetailsFn called', () => {
    it('should call shareFlightStatusDetailsFn when Share button is clicked', () => {
      const { container } = createPage({ isWebView: true, shareFlightStatus: true });

      expect(container.querySelector('.share-button')).toBeInTheDocument();

      fireEvent.click(container.querySelector('.share-button'));

      expect(shareFlightStatusDetailsFnMock).toBeCalled();
      expect(analyticsEventHelper.raiseSatelliteEvent).toBeCalled();
    });
  });

  describe('in WebView and click on renderRefreshButton', () => {
    it('should call fetchFlightDetailsFn when renderRefreshButton button is clicked', () => {
      const { container } = createPage({ isWebView: true, shareFlightStatus: true });

      expect(container.querySelector('.page-header--right-button')).toBeInTheDocument();

      fireEvent.click(container.querySelector('.page-header--right-button'));

      expect(fetchFlightDetailsFnMock).toBeCalled();
    });
  });

  const createPage = (props = {}, useNullResponse = false) => {
    const FlightStatusDetailsPageResponse = useNullResponse
      ? null
      : new FlightStatusDetailsBuilder().build().flightStatusDetailsPage;
    const defaultProps = {
      flightStatusDetailsPage: {
        response: FlightStatusDetailsPageResponse
      },
      query: {
        "flight-keys": "2023-11-10:AUSELP1009|2023-11-10:ELPDAL1565"
      },
      isWebView: false,
      shareFlightStatus: false,
      fetchFlightDetailsFn: fetchFlightDetailsFnMock,
      shareFlightStatusDetailsFn: shareFlightStatusDetailsFnMock
    };
    const mergedProps = { ...defaultProps, ...props };

    return createComponent(FlightDetailsPage, { props: { ...mergedProps } });
  };
});
