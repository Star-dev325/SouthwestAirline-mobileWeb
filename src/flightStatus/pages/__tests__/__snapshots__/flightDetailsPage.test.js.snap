// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flight Details Page should pass correct date, from and to props to component SearchFlightsSummaryHeader 1`] = `
<div>
  <div
    class="flight-details"
  >
    <div>
      <div
        class="page-header"
        id=""
      >
        <span
          class="page-title"
        >
          Flight Status
        </span>
        <span
          class="page-title-details"
        >
          Flight 1632 / 1781
        </span>
      </div>
      <div />
    </div>
    <div
      class="flight-details-content"
    >
      <div
        class="flight-search-summary"
      >
        <h5
          class="page header attached flight-search-summary"
        >
          <p
            class="sub header date"
          >
            SHARED__FLIGHTS_SUMMARY_HEADER__TODAY -  Sun, Apr 9, 2017
          </p>
          <p
            class="sub header"
          >
            From: 
            Austin, TX (AUS)
          </p>
          <p
            class="sub header to"
          >
            To: 
            Atlanta, GA (ATL)
          </p>
        </h5>
      </div>
      <div
        class="results custom-container"
      >
        <div
          class="segment-details"
        >
          <div
            class="leg-details"
          >
            <div
              class="banner boarding-header"
            >
              <div
                class="banner-name"
              >
                FLIGHT_STATUS__NOW_BOARDING
              </div>
            </div>
            <div
              class="custom-container"
            >
              <div
                class="aircraft-info"
              >
                <div
                  class="flight-info-col"
                >
                  <div
                    class="flight-number"
                  >
                    <span
                      class="flight-number--label"
                    >
                      Flight
                    </span>
                    <span
                      class="flight-number--info"
                    >
                      205
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="flight-info-row flight-info"
              >
                <div
                  class="flight-info-col"
                >
                  <div
                    class="flight-info--label"
                  >
                    Departs
                  </div>
                  <div
                    class="flight-info--flight-code"
                  >
                    ATL
                  </div>
                </div>
                <div
                  class="flight-info-col"
                >
                  <div
                    class="flight-info--label"
                  >
                    Arrives
                  </div>
                  <div
                    class="flight-info--flight-code"
                  >
                    HOU
                  </div>
                </div>
              </div>
              <div
                class="segment-status"
              >
                <div
                  class="segment-status-col departure-status"
                >
                  <div
                    class="flight-time-status flight-time-status--on-time"
                  >
                    <div>
                      <span
                        class="segment-status-col flight-status positive"
                      >
                        <i
                          class="icon icon_home-flight-status"
                        />
                        <span
                          class="flight-status--text"
                        >
                          ON TIME
                        </span>
                      </span>
                    </div>
                    <div
                      class="flight-time"
                    >
                      <div
                        class="flight-time--time"
                      >
                        <span>
                          8:55
                        </span>
                        <span
                          class="time-period"
                        >
                          AM
                        </span>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="segment-status-col airplane-wrapper"
                >
                  <i
                    class="icon icon_airplane"
                  />
                </div>
                <div
                  class="segment-status-col"
                >
                  <div
                    class="flight-time-status flight-time-status--departed"
                  >
                    <div>
                      <span
                        class="segment-status-col flight-status positive"
                      >
                        <i
                          class="icon icon_home-flight-status"
                        />
                        <span
                          class="flight-status--text"
                        >
                          DEPARTED
                        </span>
                      </span>
                    </div>
                    <div
                      class="flight-time"
                    >
                      <div
                        class="flight-time--time"
                      >
                        <span>
                          1:55
                        </span>
                        <span
                          class="time-period"
                        >
                          PM
                        </span>
                      </div>
                      <div
                        class="multi-day-indicator multi-day-indicator--nextday-green"
                        data-qa="next-day-indicator"
                      >
                        <i
                          class="multi-day-indicator--icon multi-day-indicator--icon-small icon icon_next-day-indicator"
                        />
                        <p
                          class="multi-day-indicator--font-small"
                        >
                          AIR_BOOKING__SHOPPING__NEXT_DAY
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="original-time-information"
              >
                <div
                  class="flight-info-row flight-info"
                >
                  <div
                    class="flight-info-col"
                  >
                    <div
                      class="flight-info--label"
                    >
                      Original
                    </div>
                    <div
                      class="formatted-time"
                    >
                      6:20
                      <span
                        class="formatted-time--period"
                      >
                        AM
                      </span>
                    </div>
                  </div>
                  <div
                    class="flight-info-col"
                  >
                    <div
                      class="flight-info--label"
                    >
                      Original
                    </div>
                    <div
                      class="formatted-time"
                    >
                      7:20
                      <span
                        class="formatted-time--period"
                      >
                        AM
                      </span>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="gate-information"
              >
                <div
                  class="flight-info-row flight-info"
                >
                  <div
                    class="flight-info-col"
                  >
                    <div
                      class="flight-info--label"
                    >
                      Gate
                    </div>
                    <div
                      class="flight-info--gate"
                    >
                      G21
                    </div>
                  </div>
                  <div
                    class="flight-info-col"
                  >
                    <div
                      class="flight-info--label"
                    >
                      Gate
                    </div>
                    <div
                      class="flight-info--gate"
                    >
                      N/A
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
